src/main/java/com/salesforce/storm/spout/dynamic/MessageBus.java

```diff
@@ -0,0 +1,177 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.salesforce.storm.spout.dynamic;
+
+import com.salesforce.storm.spout.dynamic.buffer.MessageBuffer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * Provides an abstraction for passing messages between the DynamicSpout instance
+ * and VirtualSpout instances.
+ *
+ * Since these instances live across different threads internally we make use of
+ * thread safe concurrent data structures making this instance ThreadSafe.
+ */
+public class MessageBus implements VirtualSpoutMessageBus, SpoutMessageBus {
+    private static final Logger logger = LoggerFactory.getLogger(MessageBus.class);
+
+    /**
+     * Queue for tuples that are ready to be emitted out into the topology.
+     */
+    private final MessageBuffer messageBuffer;
+
+    /**
+     * Buffer by spout consumer id of messages that have been acked.
+     */
+    private final Map<VirtualSpoutIdentifier, Queue<MessageId>> ackedTuplesQueue = new ConcurrentHashMap<>();
+
+    /**
+     * Buffer by spout consumer id of messages that have been failed.
+     */
+    private final Map<VirtualSpoutIdentifier, Queue<MessageId>> failedTuplesQueue = new ConcurrentHashMap<>();
+
+    /**
+     * Buffer for errors that need to be reported.
+     */
+    private final Queue<Throwable> reportedErrorsQueue = new ConcurrentLinkedQueue<>();
+
+    /**
+     * Constructor.
+     * @param messageBuffer Provides the MessageBuffer implementation to use.
+     */
+    public MessageBus(final MessageBuffer messageBuffer) {
+        this.messageBuffer = messageBuffer;
+    }
+
+    // SpoutMessageBus Interface methods.
+
+    @Override
+    public Optional<Throwable> getErrors() {
+        // Poll is non-blocking.
+        return Optional.ofNullable(reportedErrorsQueue.poll());
+    }
+
+    @Override
+    public Optional<Message> nextMessage() {
+        return Optional.ofNullable(messageBuffer.poll());
+    }
+
+    @Override
+    public void ack(final MessageId id) {
+        // TODO this actually isn't thread safe :(
+        if (!ackedTuplesQueue.containsKey(id.getSrcVirtualSpoutId())) {
+            logger.warn("Acking tuple for unknown consumer");
+            return;
+        }
+
+        ackedTuplesQueue.get(id.getSrcVirtualSpoutId()).add(id);
+    }
+
+    @Override
+    public void fail(final MessageId id) {
+        // TODO this actually isn't thread safe :(
```

Comment:
Same with this one, I think this is thread safe unless you're seeing something I am not.